packages/lesswrong/lib/analyticsEvents/ultraFeedEvents.ts

```diff
@@ -0,0 +1,130 @@
+  documentId: string;
+  documentType: 'post' | 'comment' | 'spotlight';
+  postId?: string;
+  enabled?: boolean;
+}
+
+const VIEW_THRESHOLD_MS = 500; // Minimum time on screen to trigger 'viewed'
+
+export const useUltraFeedLogging = ({
+  documentId,
+  documentType,
+  postId,
+  enabled = true,
+}: UseUltraFeedLoggingOptions) => {
+  const [logEventMutation] = useMutation(LOG_ULTRA_FEED_EVENT);
+
+  const targetRef = useRef<HTMLDivElement | null>(null);
+  const viewedSentRef = useRef(false);
+  const timerRef = useRef<NodeJS.Timeout | null>(null);
+
+  const cleanUpObserver = useRef<() => void>(() => {});
+
+  useEffect(() => {
+    const currentElement = targetRef.current as Element | null;
+    if (!enabled || !currentElement || viewedSentRef.current) {
+      return;
+    }
+
+    const observer = new IntersectionObserver(
+      (entries) => {
+        entries.forEach((entry) => {
+          if (entry.isIntersecting) {
+            if (!timerRef.current) {
+              timerRef.current = setTimeout(() => {
+                if (!viewedSentRef.current) {
+                  const eventInput: UltraFeedEventInput = {
+                    eventType: 'viewed',
+                    documentId,
+                    documentType,
+                    ...(postId && { postId }),
+                  };
+                  logEventMutation({ variables: { input: eventInput } }).catch(err => {
+                    // eslint-disable-next-line no-console
+                    console.log("Failed to log UltraFeed 'viewed' event:", err);
+                  });
+
+                  viewedSentRef.current = true;
+                  observer.disconnect();
+                }
+                timerRef.current = null;
+              }, VIEW_THRESHOLD_MS);
+            }
+          } else {
+            if (timerRef.current) {
+              clearTimeout(timerRef.current);
+              timerRef.current = null;
+            }
+          }
+        });
+      },
+      {
+        root: null,
+        threshold: 0.5,
+      }
+    );
+
+    // Observe the element
+    observer.observe(currentElement);
+
+    cleanUpObserver.current = () => {
+      observer.disconnect();
+      if (timerRef.current) {
+        clearTimeout(timerRef.current);
+        timerRef.current = null;
+      }
+    };
+
+    return () => {
+      cleanUpObserver.current();
+    };
+  }, [enabled, documentId, documentType, postId, logEventMutation]);
+
+  const trackExpansion = useCallback(
+    (level: number, maxLevelReached: boolean, wordCount: number) => {
+      if (!enabled) return;
+      const eventInput: UltraFeedEventInput = {
+        eventType: 'expanded',
+        documentId,
+        documentType: documentType as 'post' | 'comment',
+        ...(postId && { postId }),
+        expansionLevel: level,
+        maxExpansionReached: maxLevelReached,
+        wordCount,
+      };
+      logEventMutation({ variables: { input: eventInput } }).catch(err => {
```

Comment:
I've seen a lot of `.catch` in this PR when I'm not even sure that these mutations throw errors (as opposed to having a second `options` argument in `useMutation`, which accepts an `onError`)